### PR TITLE
[rllib] Rough port of DQN to build_tf_policy() pattern

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

This is a straightforward port of DQN to the builder pattern. We can clean up the code in the future more with further refactoring.

## Related issue number

https://github.com/ray-project/ray/issues/4822

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

**Comment by richardliaw** [comment]
Can you please assign this to someone else? Happy to review in general, just prefer not this week.

**Comment by ericl** [comment]
Ok, @joneswong could you take a look instead?

**Comment by joneswong** [comment]
> Ok, @joneswong could you take a look instead?

All right. I am reading #4795 in order to follow up

**Comment by ericl** [comment]
Ping on review here.

**Comment by ericl** [comment]
Ping on review here. Let me know if don't have the time.

**Comment by joneswong** [comment]
> Ping on review here. Let me know if don't have the time.

@ericl  Hi Eric, I will receive an important promotion interview next week. Please assign to other committers this time. I will be back after that.

**Comment by ericl** [comment]
Friendly ping on review.

File: python/ray/rllib/agents/dqn/dqn_policy.py [hunk 1]
```diff
@@ -706,3 +683,27 @@ def _scope_vars(scope, trainable_only=False):
...
+    gradients_fn=clip_gradients,
+    extra_action_feed_fn=exploration_setting_inputs,
+    extra_action_fetches_fn=lambda policy: {"q_values": policy.q_values},
+    extra_learn_fetches_fn=lambda policy: {"td_error": policy.loss.td_error},
+    update_ops_fn=lambda policy: policy.q_batchnorm_update_ops,
+    before_init=setup_early_mixins,
+    after_init=setup_late_mixins,
+    obs_include_prev_action_reward=False,
+    mixins=[
+        ExplorationStateMixin,
```
Comment (richardliaw) [inline python/ray/rllib/agents/dqn/dqn_policy.py]: can you remind me why the mixins need to go in at this list and also at `before_init/after_init`?

**Review by richardliaw — COMMENTED** [review]
Overall looks fine; just a few questions for clarity

**Review by richardliaw — APPROVED** [review]

File: python/ray/rllib/policy/tf_policy.py [hunk 2]
```diff
@@ -139,6 +139,39 @@ def __init__(self,
...
+            raise RuntimeError(
+                "You cannot call policy.get_placeholder() for non-obs inputs "
+                "before the loss has been initialized. To avoid this, use "
+                "policy.loss_initialized() to check whether this is the "
+                "case, or move the call to later (e.g., from stats_fn to "
+                "grad_stats_fn).")
+
+        return self._loss_input_dict[name]
+
+    def get_session(self):
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/policy/tf_policy.py]: Acknowledged, won't fix.

File: python/ray/rllib/evaluation/tf_policy_template.py [hunk 3]
```diff
@@ -1,146 +0,0 @@
-from __future__ import absolute_import
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/evaluation/tf_policy_template.py]: It was moved.

File: python/ray/rllib/agents/dqn/dqn_policy.py [hunk 4]
```diff
@@ -446,166 +339,250 @@ def __init__(self, observation_space, action_space, config):
...
-    @override(Policy)
-    def get_state(self):
-        return [TFPolicy.get_state(self), self.cur_epsilon]
-
-    @override(Policy)
-    def set_state(self, state):
-        TFPolicy.set_state(self, state[0])
-        self.set_epsilon(state[1])
+    def update_target(self):
+        return self.get_session().run(self.update_target_expr)
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/agents/dqn/dqn_policy.py]: Acknowledged, won't fix.

File: python/ray/rllib/agents/dqn/dqn_policy.py (hunk 1 above)
Comment (ericl) [reply python/ray/rllib/agents/dqn/dqn_policy.py]: This is for the class declaration. The constructor needs to be called separately, it's not possible to figure out where to call it just based on the class name.

**Pull request merged**